runtime/stdlib/account.go

```diff
@@ -0,0 +1,1718 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package stdlib
+
+import (
+	"fmt"
+
+	"golang.org/x/crypto/sha3"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/parser"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+const authAccountFunctionDocString = `
+Creates a new account, paid by the given existing account
+`
+
+var authAccountFunctionType = &sema.FunctionType{
+	Parameters: []*sema.Parameter{
+		{
+			Identifier: "payer",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.AuthAccountType,
+			),
+		},
+	},
+	ReturnTypeAnnotation: sema.NewTypeAnnotation(
+		sema.AuthAccountType,
+	),
+}
+
+type EventEmitter interface {
+	EmitEvent(
+		inter *interpreter.Interpreter,
+		eventType *sema.CompositeType,
+		values []interpreter.Value,
+		getLocationRange func() interpreter.LocationRange,
+	)
+}
+
+type AuthAccountHandler interface {
+	BalanceProvider
+	AvailableBalanceProvider
+	StorageUsedProvider
+	StorageCapacityProvider
+	AccountEncodedKeyAdditionHandler
+	AccountEncodedKeyRevocationHandler
+	AuthAccountKeysHandler
+	AuthAccountContractsHandler
+}
+
+type AccountCreator interface {
+	EventEmitter
+	AuthAccountHandler
+	// CreateAccount creates a new account.
+	CreateAccount(payer common.Address) (address common.Address, err error)
+}
+
+func NewAuthAccountConstructor(creator AccountCreator) StandardLibraryValue {
```

Comment:
Each standard library value declaration now only depends on a dedicated interface. This allows the declarations to be used independent of the runtime package, and makes them individually testable.

The environment implements these interfaces and forwards the calls to the current `runtime.Interface` of the environment (the environment is reused across multiple script and transaction invocations, but its interface is swapped)

Reply:
I assume the code was just 'moved' and updated to rely on the new interface, with no change to the logic?

Reply:
Yes, I only updated the "external calls", the code was mostly just moved over. We can maybe go over it side-by-side in a review session